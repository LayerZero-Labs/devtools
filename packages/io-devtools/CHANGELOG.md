# @layerzerolabs/io-devtools

## 0.1.10

### Patch Changes

- 8b32a79: Use less strict version requirements for layerzero peer dependencies

## 0.1.9

### Patch Changes

- 938ac3d: Improve error reporting on UIntBigIntSchema

## 0.1.8

### Patch Changes

- 9af95ca: Add createWithAsyncLogger logging utility for async functions

## 0.1.7

### Patch Changes

- 8a4addf: Replace cli-table3 with table

## 0.1.6

### Patch Changes

- 2a36974: Add the ability to define configs as functions

## 0.1.5

### Patch Changes

- 54cf16e: Silence bigint-buffer warning

## 0.1.4

### Patch Changes

- 9e78abe: Update hardhat-deploy

## 0.1.3

### Patch Changes

- e766c86: Adding in Default Colors to print outs

## 0.1.2

### Patch Changes

- 253c79e: Rename description to hint for select prompts
- 5c58d69: Update @layerzerolabs dependencies to 2.1.3

## 0.1.1

### Patch Changes

- 40e8dba: Remove CommaSeparatedValuesSchema in favor of splitCommaSeparated; Move LogLevelSchema; Export isLogLevel

## 0.1.0

### Minor Changes

- 120adf1: Make packages public

## 0.0.7

### Patch Changes

- 81d7eee: Add promptForText utility

## 0.0.6

### Patch Changes

- 2b36ed9: Updating Table for Get OApp Config Task
- 2b36ed9: Rafactor some of the printers in io-devtools

## 0.0.5

### Patch Changes

- ed450c4: Add promptToSelectOne and promptToSelectMultiple utilities

## 0.0.4

### Patch Changes

- af8cc25: Update dependencies

## 0.0.3

### Patch Changes

- 5077bf1: Unmount the component when using printTable

## 0.0.2

### Patch Changes

- e0be8b7: Reexport Logger type
- 70646b4: Add missing dependencies

## 0.0.1

### Patch Changes

- b74afbe: Initial 0.0.1 version
