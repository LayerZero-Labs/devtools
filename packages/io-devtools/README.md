<p align="center">
  <a href="https://layerzero.network">
    <img alt="LayerZero" style="max-width: 500px" src="https://d3a2dpnnrypp5h.cloudfront.net/bridge-app/lz.png"/>
  </a>
</p>

<h1 align="center">@layerzerolabs/io-utils</h1>

<!-- The badges section -->
<p align="center">
  <!-- Shields.io NPM published package version -->
  <a href="https://www.npmjs.com/package/@layerzerolabs/io-utils"><img alt="NPM Version" src="https://img.shields.io/npm/v/@layerzerolabs/io-utils"/></a>
  <!-- Shields.io NPM downloads -->
  <a href="https://www.npmjs.com/package/@layerzerolabs/io-utils"><img alt="Downloads" src="https://img.shields.io/npm/dm/@layerzerolabs/io-utils"/></a>
  <!-- Shields.io license badge -->
  <a href="https://www.npmjs.com/package/@layerzerolabs/io-utils"><img alt="NPM License" src="https://img.shields.io/npm/l/@layerzerolabs/io-utils"/></a>
</p>

## Installation

```bash
pnpm install @layerzerolabs/io-utils
```

```bash
pnpm install @layerzerolabs/io-utils
```

```bash
npm install @layerzerolabs/io-utils
```

## API Documentation

### Filesystem utilities

#### isDirectory(path)

Returns `true` if specified filesystem `path` points to a directory, `false` otherwise. Does not throw if `path` does not exist on the filesystem, instead returns `false`

#### isFile(path)

Returns `true` if specified filesystem `path` points to a file, `false` otherwise. Does not throw if `path` does not exist on the filesystem, instead returns `false`

#### isReadable(path)

Returns `true` if specified filesystem `path` can be read by the current user, `false` otherwise. Does not throw if `path` does not exist on the filesystem, instead returns `false`

### Standard input/output utilities

#### promptToContinue([message, defaultValue])

Asks the user whether they want to continue and reads the input from the CLI standard input. By default the question displayed is `Do you want to continue?` and the default response is `yes`

```typescript
const goahead = await promptToContinue();

// To ask a different question
const goahead = await promptToContinue("Are you sure?");

// To default the response to false, good for important and unsafe decisions
const goahead = await promptToContinue("Are you sure?", false);
```
