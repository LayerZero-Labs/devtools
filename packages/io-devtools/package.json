{
  "name": "@layerzerolabs/io-devtools",
  "version": "0.1.10",
  "description": "Utilities for working with I/O in LayerZero devtools",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/LayerZero-Labs/devtools.git",
    "directory": "packages/io-devtools"
  },
  "license": "MIT",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "require": "./dist/index.js",
      "import": "./dist/index.mjs"
    },
    "./*": {
      "types": "./dist/*/index.d.ts",
      "require": "./dist/*/index.js",
      "import": "./dist/*/index.mjs"
    }
  },
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "files": [
    "dist",
    "swag"
  ],
  "scripts": {
    "prebuild": "tsc -noEmit",
    "build": "$npm_execpath tsup",
    "clean": "rm -rf dist",
    "dev": "$npm_execpath tsup --watch",
    "lint": "$npm_execpath eslint '**/*.{js,ts,json}'",
    "lint:fix": "eslint --fix '**/*.{js,ts,json}'",
    "test": "jest --ci"
  },
  "dependencies": {
    "chalk": "^4.1.2",
    "logform": "^2.6.0",
    "prompts": "^2.4.2",
    "table": "~6.8.2",
    "winston": "^3.11.0"
  },
  "devDependencies": {
    "@swc/core": "^1.4.0",
    "@swc/jest": "^0.2.36",
    "@types/ink-gradient": "^2.0.4",
    "@types/jest": "^29.5.12",
    "@types/prompts": "^2.4.9",
    "@types/react": "^17.0.75",
    "fast-check": "^3.15.1",
    "ink": "^3.2.0",
    "ink-gradient": "^2.0.0",
    "ink-table": "^3.1.0",
    "jest": "^29.7.0",
    "jest-extended": "^4.0.2",
    "react": "^17.0.2",
    "ts-node": "^10.9.2",
    "tslib": "~2.6.2",
    "tsup": "~8.0.1",
    "typescript": "^5.3.3",
    "zod": "^3.22.4"
  },
  "peerDependencies": {
    "ink": "^3.2.0",
    "ink-gradient": "^2.0.0",
    "ink-table": "^3.1.0",
    "react": "^17.0.2",
    "yoga-layout-prebuilt": "^1.9.6",
    "zod": "^3.22.4"
  },
  "peerDependenciesMeta": {
    "ink": {
      "optional": true
    },
    "ink-gradient": {
      "optional": true
    },
    "ink-table": {
      "optional": true
    },
    "react": {
      "optional": true
    },
    "yoga-layout-prebuilt": {
      "optional": true
    }
  },
  "publishConfig": {
    "access": "public"
  }
}
