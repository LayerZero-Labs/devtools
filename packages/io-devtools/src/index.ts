export * from './async'
export * from './config'
export * from './filesystem'
export * from './language'
export * from './stdio'
