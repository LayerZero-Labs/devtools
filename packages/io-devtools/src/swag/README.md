# `@layerzerolabs/io-devtools/swag`

This directory is not included in the default bundle (`@layerzerolabs/io-devtools`) because it introduces dependencies that are not required for the basic functionality and that would bloat the bundle size.

If you need to use this module, you will need to include the `peerDependencies` that are listed as optional under `peerDependenciesMeta` in `package.json`
