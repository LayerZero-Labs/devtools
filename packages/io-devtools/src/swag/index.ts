export * from './printer'
export * from './renderer'
