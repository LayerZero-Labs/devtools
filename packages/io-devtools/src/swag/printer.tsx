import React from "react";
import { render } from "ink";
import { Logo } from "./components/logo";
import { RecordList, RecordData, Record } from "./components/record";

export const printLogo = () => render(<Logo />).unmount();

export const printRecord = (data: RecordData) =>
  render(<Record data={data} />).unmount();

export const printRecords = (data: RecordData[]) =>
  render(<RecordList data={data} />).unmount();
