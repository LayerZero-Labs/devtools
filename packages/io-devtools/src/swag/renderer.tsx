import React from "react";
import { ProgressBar, ProgressBarProps } from "./components/progress";

/**
 * Creates a ProgressBar React node.
 *
 * This is so that the consumers don't need to use JSX and/or React.createElement
 * and so that we can pass this node to e.g. instance.rerender() if a rendered ink component
 *
 * ```typescript
 * import { createProgressBar, render } from '@layerzerolabs/io-devtools/swag'
 *
 * const instance = render(createProgressBar())
 *
 * instance.rerender(createProgressBar({ progress: 0.5 }))
 * instance.clear()
 * ```
 *
 * @param {ProgressBarProps} props
 * @returns {React.ReactElement<ProgressBarProps>}
 */
export const createProgressBar = (
  props: ProgressBarProps = {},
): React.ReactElement<ProgressBarProps> => <ProgressBar {...props} />;

/**
 * Re-export the render from ink so that we can use it without importing directly from ink
 */
export { render } from "ink";
