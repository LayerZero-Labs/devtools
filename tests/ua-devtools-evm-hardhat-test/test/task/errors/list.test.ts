import hre from 'hardhat'
import { TASK_COMPILE } from 'hardhat/builtin-tasks/task-names'
import { TASK_LZ_ERRORS_LIST } from '@layerzerolabs/ua-devtools-evm-hardhat'
import { printVerticalTable } from '@layerzerolabs/io-devtools/swag'

jest.mock('@layerzerolabs/io-devtools/swag', () => ({
    printVerticalTable: jest.fn(),
    printLogo: jest.fn(),
}))

const printTableMock = printVerticalTable as jest.Mock
const runMock = jest.spyOn(hre, 'run')

describe(`task ${TASK_LZ_ERRORS_LIST}`, () => {
    beforeEach(() => {
        printTableMock.mockReset()
        runMock.mockClear()
    })

    it('should compile contracts', async () => {
        await hre.run(TASK_LZ_ERRORS_LIST, {})

        // For some reason even though we did not specify any arguments to the compile task,
        // jest still sees some aarguments being passed so we need to pass those to make this expect work
        expect(runMock).toHaveBeenCalledWith(TASK_COMPILE, undefined, {}, undefined)
    })

    it('should not print anything if there are no matching errors', async () => {
        await hre.run(TASK_LZ_ERRORS_LIST, { containing: 'idontexist' })

        expect(printTableMock).not.toHaveBeenCalled()
    })

    it('should print all matching errors if `containing` argument is not supplied', async () => {
        await hre.run(TASK_LZ_ERRORS_LIST, { containing: undefined })

        expect(printTableMock).toHaveBeenCalledTimes(1)
        expect(printTableMock.mock.calls[0]).toMatchSnapshot()
    })

    it('should print all errors with matching name if `containing` argument is supplied', async () => {
        await hre.run(TASK_LZ_ERRORS_LIST, { containing: 'Invalid' })

        expect(printTableMock).toHaveBeenCalledTimes(1)
        expect(printTableMock.mock.calls[0]).toMatchSnapshot()
    })

    it('should print all errors with matching signature if `containing` argument is supplied', async () => {
        await hre.run(TASK_LZ_ERRORS_LIST, { containing: '0x81da67ee' })

        expect(printTableMock).toHaveBeenCalledTimes(1)
        expect(printTableMock.mock.calls[0]).toMatchSnapshot()
    })
})
