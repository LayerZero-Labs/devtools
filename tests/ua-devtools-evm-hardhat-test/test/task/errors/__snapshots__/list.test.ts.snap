// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`task lz:errors:list should print all matching errors if \`containing\` argument is not supplied 1`] = `
[
  [
    {
      "Contract": "DVNOptions",
      "Error": "InvalidDVNIdx()",
    },
    {
      "Contract": "DVNOptions",
      "Error": "InvalidDVNOptions(uint256)",
    },
    {
      "Contract": "OApp",
      "Error": "InvalidEndpointCall()",
    },
    {
      "Contract": "OApp",
      "Error": "LzTokenUnavailable()",
    },
    {
      "Contract": "OApp",
      "Error": "NoPeer(uint32)",
    },
    {
      "Contract": "OApp",
      "Error": "NotEnoughNative(uint256)",
    },
    {
      "Contract": "OApp",
      "Error": "OnlyEndpoint(address)",
    },
    {
      "Contract": "OApp",
      "Error": "OnlyPeer(uint32,bytes32)",
    },
    {
      "Contract": "OAppCore",
      "Error": "InvalidEndpointCall()",
    },
    {
      "Contract": "OAppCore",
      "Error": "NoPeer(uint32)",
    },
    {
      "Contract": "OAppCore",
      "Error": "OnlyPeer(uint32,bytes32)",
    },
    {
      "Contract": "OAppReceiver",
      "Error": "InvalidEndpointCall()",
    },
    {
      "Contract": "OAppReceiver",
      "Error": "NoPeer(uint32)",
    },
    {
      "Contract": "OAppReceiver",
      "Error": "OnlyEndpoint(address)",
    },
    {
      "Contract": "OAppReceiver",
      "Error": "OnlyPeer(uint32,bytes32)",
    },
    {
      "Contract": "OAppSender",
      "Error": "InvalidEndpointCall()",
    },
    {
      "Contract": "OAppSender",
      "Error": "LzTokenUnavailable()",
    },
    {
      "Contract": "OAppSender",
      "Error": "NoPeer(uint32)",
    },
    {
      "Contract": "OAppSender",
      "Error": "NotEnoughNative(uint256)",
    },
    {
      "Contract": "OAppSender",
      "Error": "OnlyPeer(uint32,bytes32)",
    },
    {
      "Contract": "OmniCounter",
      "Error": "InvalidEndpointCall()",
    },
    {
      "Contract": "OmniCounter",
      "Error": "InvalidOptionType(uint16)",
    },
    {
      "Contract": "OmniCounter",
      "Error": "LzTokenUnavailable()",
    },
    {
      "Contract": "OmniCounter",
      "Error": "NoPeer(uint32)",
    },
    {
      "Contract": "OmniCounter",
      "Error": "NotEnoughNative(uint256)",
    },
    {
      "Contract": "OmniCounter",
      "Error": "OnlyEndpoint(address)",
    },
    {
      "Contract": "OmniCounter",
      "Error": "OnlyPeer(uint32,bytes32)",
    },
    {
      "Contract": "OmniCounter",
      "Error": "OnlySelf()",
    },
    {
      "Contract": "OmniCounter",
      "Error": "SimulationResult(bytes)",
    },
    {
      "Contract": "IOAppCore",
      "Error": "InvalidEndpointCall()",
    },
    {
      "Contract": "IOAppCore",
      "Error": "NoPeer(uint32)",
    },
    {
      "Contract": "IOAppCore",
      "Error": "OnlyPeer(uint32,bytes32)",
    },
    {
      "Contract": "IOAppOptionsType3",
      "Error": "InvalidOptions(bytes)",
    },
    {
      "Contract": "OAppOptionsType3",
      "Error": "InvalidOptions(bytes)",
    },
    {
      "Contract": "OptionsBuilder",
      "Error": "InvalidOptionType(uint16)",
    },
    {
      "Contract": "OptionsBuilder",
      "Error": "InvalidSize(uint256,uint256)",
    },
    {
      "Contract": "OAppPreCrimeSimulator",
      "Error": "OnlySelf()",
    },
    {
      "Contract": "OAppPreCrimeSimulator",
      "Error": "SimulationResult(bytes)",
    },
    {
      "Contract": "IOAppPreCrimeSimulator",
      "Error": "OnlySelf()",
    },
    {
      "Contract": "IOAppPreCrimeSimulator",
      "Error": "SimulationResult(bytes)",
    },
    {
      "Contract": "IPreCrime",
      "Error": "CrimeFound(bytes)",
    },
    {
      "Contract": "IPreCrime",
      "Error": "InvalidSimulationResult(uint32,bytes)",
    },
    {
      "Contract": "IPreCrime",
      "Error": "OnlyOffChain()",
    },
    {
      "Contract": "IPreCrime",
      "Error": "PacketOversize(uint256,uint256)",
    },
    {
      "Contract": "IPreCrime",
      "Error": "PacketUnsorted()",
    },
    {
      "Contract": "IPreCrime",
      "Error": "SimulationFailed(bytes)",
    },
    {
      "Contract": "IPreCrime",
      "Error": "SimulationResultNotFound(uint32)",
    },
    {
      "Contract": "Errors",
      "Error": "AlreadyRegistered()",
    },
    {
      "Contract": "Errors",
      "Error": "ComposeExists()",
    },
    {
      "Contract": "Errors",
      "Error": "ComposeNotFound(bytes32,bytes32)",
    },
    {
      "Contract": "Errors",
      "Error": "DefaultReceiveLibUnavailable()",
    },
    {
      "Contract": "Errors",
      "Error": "DefaultSendLibUnavailable()",
    },
    {
      "Contract": "Errors",
      "Error": "InsufficientFee(uint256,uint256,uint256,uint256)",
    },
    {
      "Contract": "Errors",
      "Error": "InvalidAddress()",
    },
    {
      "Contract": "Errors",
      "Error": "InvalidAmount(uint256,uint256)",
    },
    {
      "Contract": "Errors",
      "Error": "InvalidArgument()",
    },
    {
      "Contract": "Errors",
      "Error": "InvalidExpiry()",
    },
    {
      "Contract": "Errors",
      "Error": "InvalidNonce(uint64)",
    },
    {
      "Contract": "Errors",
      "Error": "InvalidPayloadHash()",
    },
    {
      "Contract": "Errors",
      "Error": "InvalidReceiveLibrary()",
    },
    {
      "Contract": "Errors",
      "Error": "InvalidSizeForAddress()",
    },
    {
      "Contract": "Errors",
      "Error": "LzTokenUnavailable()",
    },
    {
      "Contract": "Errors",
      "Error": "NotImplemented()",
    },
    {
      "Contract": "Errors",
      "Error": "OnlyAltToken()",
    },
    {
      "Contract": "Errors",
      "Error": "OnlyNonDefaultLib()",
    },
    {
      "Contract": "Errors",
      "Error": "OnlyReceiveLib()",
    },
    {
      "Contract": "Errors",
      "Error": "OnlyRegisteredLib()",
    },
    {
      "Contract": "Errors",
      "Error": "OnlyRegisteredOrDefaultLib()",
    },
    {
      "Contract": "Errors",
      "Error": "OnlySendLib()",
    },
    {
      "Contract": "Errors",
      "Error": "PathNotInitializable()",
    },
    {
      "Contract": "Errors",
      "Error": "PathNotVerifiable()",
    },
    {
      "Contract": "Errors",
      "Error": "PayloadHashNotFound(bytes32,bytes32)",
    },
    {
      "Contract": "Errors",
      "Error": "SameValue()",
    },
    {
      "Contract": "Errors",
      "Error": "SendReentrancy()",
    },
    {
      "Contract": "Errors",
      "Error": "Unauthorized()",
    },
    {
      "Contract": "Errors",
      "Error": "UnsupportedEid()",
    },
    {
      "Contract": "Errors",
      "Error": "UnsupportedInterface()",
    },
    {
      "Contract": "Errors",
      "Error": "ZeroLzTokenFee()",
    },
    {
      "Contract": "ExecutorOptions",
      "Error": "InvalidLzComposeOption()",
    },
    {
      "Contract": "ExecutorOptions",
      "Error": "InvalidLzReceiveOption()",
    },
    {
      "Contract": "ExecutorOptions",
      "Error": "InvalidNativeDropOption()",
    },
    {
      "Contract": "DefaultOApp",
      "Error": "InvalidEndpointCall()",
    },
    {
      "Contract": "DefaultOApp",
      "Error": "InvalidOptions(bytes)",
    },
    {
      "Contract": "DefaultOApp",
      "Error": "LzTokenUnavailable()",
    },
    {
      "Contract": "DefaultOApp",
      "Error": "NoPeer(uint32)",
    },
    {
      "Contract": "DefaultOApp",
      "Error": "NotEnoughNative(uint256)",
    },
    {
      "Contract": "DefaultOApp",
      "Error": "OnlyEndpoint(address)",
    },
    {
      "Contract": "DefaultOApp",
      "Error": "OnlyPeer(uint32,bytes32)",
    },
    {
      "Contract": "OmniCounter",
      "Error": "InvalidEndpointCall()",
    },
    {
      "Contract": "OmniCounter",
      "Error": "InvalidOptionType(uint16)",
    },
    {
      "Contract": "OmniCounter",
      "Error": "LzTokenUnavailable()",
    },
    {
      "Contract": "OmniCounter",
      "Error": "NoPeer(uint32)",
    },
    {
      "Contract": "OmniCounter",
      "Error": "NotEnoughNative(uint256)",
    },
    {
      "Contract": "OmniCounter",
      "Error": "OnlyEndpoint(address)",
    },
    {
      "Contract": "OmniCounter",
      "Error": "OnlyPeer(uint32,bytes32)",
    },
    {
      "Contract": "OmniCounter",
      "Error": "OnlySelf()",
    },
    {
      "Contract": "OmniCounter",
      "Error": "SimulationResult(bytes)",
    },
  ],
]
`;

exports[`task lz:errors:list should print all matching errors if \`containing\` argument is supplied 1`] = `
[
  [
    {
      "Contract": "DVNOptions",
      "Error": "InvalidDVNIdx()",
    },
    {
      "Contract": "DVNOptions",
      "Error": "InvalidDVNOptions(uint256)",
    },
    {
      "Contract": "OApp",
      "Error": "InvalidEndpointCall()",
    },
    {
      "Contract": "OApp",
      "Error": "LzTokenUnavailable()",
    },
    {
      "Contract": "OApp",
      "Error": "NoPeer(uint32)",
    },
    {
      "Contract": "OApp",
      "Error": "NotEnoughNative(uint256)",
    },
    {
      "Contract": "OApp",
      "Error": "OnlyEndpoint(address)",
    },
    {
      "Contract": "OApp",
      "Error": "OnlyPeer(uint32,bytes32)",
    },
    {
      "Contract": "OAppCore",
      "Error": "InvalidEndpointCall()",
    },
    {
      "Contract": "OAppCore",
      "Error": "NoPeer(uint32)",
    },
    {
      "Contract": "OAppCore",
      "Error": "OnlyPeer(uint32,bytes32)",
    },
    {
      "Contract": "OAppReceiver",
      "Error": "InvalidEndpointCall()",
    },
    {
      "Contract": "OAppReceiver",
      "Error": "NoPeer(uint32)",
    },
    {
      "Contract": "OAppReceiver",
      "Error": "OnlyEndpoint(address)",
    },
    {
      "Contract": "OAppReceiver",
      "Error": "OnlyPeer(uint32,bytes32)",
    },
    {
      "Contract": "OAppSender",
      "Error": "InvalidEndpointCall()",
    },
    {
      "Contract": "OAppSender",
      "Error": "LzTokenUnavailable()",
    },
    {
      "Contract": "OAppSender",
      "Error": "NoPeer(uint32)",
    },
    {
      "Contract": "OAppSender",
      "Error": "NotEnoughNative(uint256)",
    },
    {
      "Contract": "OAppSender",
      "Error": "OnlyPeer(uint32,bytes32)",
    },
    {
      "Contract": "OmniCounter",
      "Error": "InvalidEndpointCall()",
    },
    {
      "Contract": "OmniCounter",
      "Error": "InvalidOptionType(uint16)",
    },
    {
      "Contract": "OmniCounter",
      "Error": "LzTokenUnavailable()",
    },
    {
      "Contract": "OmniCounter",
      "Error": "NoPeer(uint32)",
    },
    {
      "Contract": "OmniCounter",
      "Error": "NotEnoughNative(uint256)",
    },
    {
      "Contract": "OmniCounter",
      "Error": "OnlyEndpoint(address)",
    },
    {
      "Contract": "OmniCounter",
      "Error": "OnlyPeer(uint32,bytes32)",
    },
    {
      "Contract": "OmniCounter",
      "Error": "OnlySelf()",
    },
    {
      "Contract": "OmniCounter",
      "Error": "SimulationResult(bytes)",
    },
    {
      "Contract": "IOAppCore",
      "Error": "InvalidEndpointCall()",
    },
    {
      "Contract": "IOAppCore",
      "Error": "NoPeer(uint32)",
    },
    {
      "Contract": "IOAppCore",
      "Error": "OnlyPeer(uint32,bytes32)",
    },
    {
      "Contract": "IOAppOptionsType3",
      "Error": "InvalidOptions(bytes)",
    },
    {
      "Contract": "OAppOptionsType3",
      "Error": "InvalidOptions(bytes)",
    },
    {
      "Contract": "OptionsBuilder",
      "Error": "InvalidOptionType(uint16)",
    },
    {
      "Contract": "OptionsBuilder",
      "Error": "InvalidSize(uint256,uint256)",
    },
    {
      "Contract": "OAppPreCrimeSimulator",
      "Error": "OnlySelf()",
    },
    {
      "Contract": "OAppPreCrimeSimulator",
      "Error": "SimulationResult(bytes)",
    },
    {
      "Contract": "IOAppPreCrimeSimulator",
      "Error": "OnlySelf()",
    },
    {
      "Contract": "IOAppPreCrimeSimulator",
      "Error": "SimulationResult(bytes)",
    },
    {
      "Contract": "IPreCrime",
      "Error": "CrimeFound(bytes)",
    },
    {
      "Contract": "IPreCrime",
      "Error": "InvalidSimulationResult(uint32,bytes)",
    },
    {
      "Contract": "IPreCrime",
      "Error": "OnlyOffChain()",
    },
    {
      "Contract": "IPreCrime",
      "Error": "PacketOversize(uint256,uint256)",
    },
    {
      "Contract": "IPreCrime",
      "Error": "PacketUnsorted()",
    },
    {
      "Contract": "IPreCrime",
      "Error": "SimulationFailed(bytes)",
    },
    {
      "Contract": "IPreCrime",
      "Error": "SimulationResultNotFound(uint32)",
    },
    {
      "Contract": "Errors",
      "Error": "AlreadyRegistered()",
    },
    {
      "Contract": "Errors",
      "Error": "ComposeExists()",
    },
    {
      "Contract": "Errors",
      "Error": "ComposeNotFound(bytes32,bytes32)",
    },
    {
      "Contract": "Errors",
      "Error": "DefaultReceiveLibUnavailable()",
    },
    {
      "Contract": "Errors",
      "Error": "DefaultSendLibUnavailable()",
    },
    {
      "Contract": "Errors",
      "Error": "InsufficientFee(uint256,uint256,uint256,uint256)",
    },
    {
      "Contract": "Errors",
      "Error": "InvalidAddress()",
    },
    {
      "Contract": "Errors",
      "Error": "InvalidAmount(uint256,uint256)",
    },
    {
      "Contract": "Errors",
      "Error": "InvalidArgument()",
    },
    {
      "Contract": "Errors",
      "Error": "InvalidExpiry()",
    },
    {
      "Contract": "Errors",
      "Error": "InvalidNonce(uint64)",
    },
    {
      "Contract": "Errors",
      "Error": "InvalidPayloadHash()",
    },
    {
      "Contract": "Errors",
      "Error": "InvalidReceiveLibrary()",
    },
    {
      "Contract": "Errors",
      "Error": "InvalidSizeForAddress()",
    },
    {
      "Contract": "Errors",
      "Error": "LzTokenUnavailable()",
    },
    {
      "Contract": "Errors",
      "Error": "NotImplemented()",
    },
    {
      "Contract": "Errors",
      "Error": "OnlyAltToken()",
    },
    {
      "Contract": "Errors",
      "Error": "OnlyNonDefaultLib()",
    },
    {
      "Contract": "Errors",
      "Error": "OnlyReceiveLib()",
    },
    {
      "Contract": "Errors",
      "Error": "OnlyRegisteredLib()",
    },
    {
      "Contract": "Errors",
      "Error": "OnlyRegisteredOrDefaultLib()",
    },
    {
      "Contract": "Errors",
      "Error": "OnlySendLib()",
    },
    {
      "Contract": "Errors",
      "Error": "PathNotInitializable()",
    },
    {
      "Contract": "Errors",
      "Error": "PathNotVerifiable()",
    },
    {
      "Contract": "Errors",
      "Error": "PayloadHashNotFound(bytes32,bytes32)",
    },
    {
      "Contract": "Errors",
      "Error": "SameValue()",
    },
    {
      "Contract": "Errors",
      "Error": "SendReentrancy()",
    },
    {
      "Contract": "Errors",
      "Error": "Unauthorized()",
    },
    {
      "Contract": "Errors",
      "Error": "UnsupportedEid()",
    },
    {
      "Contract": "Errors",
      "Error": "UnsupportedInterface()",
    },
    {
      "Contract": "Errors",
      "Error": "ZeroLzTokenFee()",
    },
    {
      "Contract": "ExecutorOptions",
      "Error": "InvalidLzComposeOption()",
    },
    {
      "Contract": "ExecutorOptions",
      "Error": "InvalidLzReceiveOption()",
    },
    {
      "Contract": "ExecutorOptions",
      "Error": "InvalidNativeDropOption()",
    },
    {
      "Contract": "DefaultOApp",
      "Error": "InvalidEndpointCall()",
    },
    {
      "Contract": "DefaultOApp",
      "Error": "InvalidOptions(bytes)",
    },
    {
      "Contract": "DefaultOApp",
      "Error": "LzTokenUnavailable()",
    },
    {
      "Contract": "DefaultOApp",
      "Error": "NoPeer(uint32)",
    },
    {
      "Contract": "DefaultOApp",
      "Error": "NotEnoughNative(uint256)",
    },
    {
      "Contract": "DefaultOApp",
      "Error": "OnlyEndpoint(address)",
    },
    {
      "Contract": "DefaultOApp",
      "Error": "OnlyPeer(uint32,bytes32)",
    },
    {
      "Contract": "OmniCounter",
      "Error": "InvalidEndpointCall()",
    },
    {
      "Contract": "OmniCounter",
      "Error": "InvalidOptionType(uint16)",
    },
    {
      "Contract": "OmniCounter",
      "Error": "LzTokenUnavailable()",
    },
    {
      "Contract": "OmniCounter",
      "Error": "NoPeer(uint32)",
    },
    {
      "Contract": "OmniCounter",
      "Error": "NotEnoughNative(uint256)",
    },
    {
      "Contract": "OmniCounter",
      "Error": "OnlyEndpoint(address)",
    },
    {
      "Contract": "OmniCounter",
      "Error": "OnlyPeer(uint32,bytes32)",
    },
    {
      "Contract": "OmniCounter",
      "Error": "OnlySelf()",
    },
    {
      "Contract": "OmniCounter",
      "Error": "SimulationResult(bytes)",
    },
  ],
]
`;
